feat: set collections for a library component [FC-0062]

navinkarkera opened the pull request:
## Description

Adds api to set/update collections on a given component. Related to https://github.com/openedx/frontend-app-authoring/pull/1373


## Supporting information

- `Private-ref`: https://tasks.opencraft.com/browse/FAL-3848
- https://github.com/openedx/frontend-app-authoring/pull/1373
- https://github.com/openedx/openedx-learning/pull/238

**Depends on:**

* https://github.com/openedx/openedx-events/pull/404

## Testing instructions

See https://github.com/openedx/frontend-app-authoring/pull/1373 for instructions.

## Deadline

"None" if there's no rush, or provide a specific date or event (and reason) if there is one.

## Other information

navinkarkera committed:
feat: add & remove collections to component

navinkarkera committed:
feat: add & remove collections to component

navinkarkera committed:
refactor: set collections in a component

navinkarkera committed:
refactor: set collections in a component

navinkarkera committed:
refactor: async collections indexing

navinkarkera committed:
refactor: async collections indexing

navinkarkera committed:
fix: lint issues

navinkarkera committed:
fix: lint issues

navinkarkera committed:
test: test set collections for component

navinkarkera committed:
test: test set collections for component

openedx-webhooks commented:
Thanks for the pull request, @navinkarkera!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

navinkarkera committed:
chore: fix docs

navinkarkera committed:
chore: fix docs

navinkarkera committed:
chore: temporarily update openedx-learning package

navinkarkera committed:
chore: temporarily update openedx-learning package

navinkarkera commented on requirements/edx/base.txt:
```diff
@@ -824,7 +824,7 @@ openedx-filters==1.10.0
     #   lti-consumer-xblock
     #   ora2
-openedx-learning==0.13.1
+openedx-learning @ git+https://github.com/open-craft/openedx-learning@navin/component-collection-api
```
`TODO`: update this after merging https://github.com/openedx/openedx-learning/pull/238 releasing new version of openedx-learning

navinkarkera commented on requirements/edx/development.txt:
```diff
@@ -1373,7 +1373,7 @@ openedx-filters==1.10.0
     #   lti-consumer-xblock
     #   ora2
-openedx-learning==0.13.1
+openedx-learning @ git+https://github.com/open-craft/openedx-learning@navin/component-collection-api
```
`TODO`: update this after merging https://github.com/openedx/openedx-learning/pull/238 releasing new version of openedx-learning

navinkarkera commented on requirements/edx/doc.txt:
```diff
@@ -983,7 +983,7 @@ openedx-filters==1.10.0
     #   lti-consumer-xblock
     #   ora2
-openedx-learning==0.13.1
+openedx-learning @ git+https://github.com/open-craft/openedx-learning@navin/component-collection-api
```
`TODO`: update this after merging https://github.com/openedx/openedx-learning/pull/238 releasing new version of openedx-learning

navinkarkera commented on requirements/edx/testing.txt:
```diff
@@ -1034,7 +1034,7 @@ openedx-filters==1.10.0
     #   lti-consumer-xblock
     #   ora2
-openedx-learning==0.13.1
+openedx-learning @ git+https://github.com/open-craft/openedx-learning@navin/component-collection-api
```
`TODO`: update this after merging https://github.com/openedx/openedx-learning/pull/238 releasing new version of openedx-learning

navinkarkera committed:
fix: lint issues

navinkarkera committed:
fix: lint issues

navinkarkera commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1235,6 +1235,53 @@ def update_library_collection_components(
+
+    from ..content.search.tasks import update_library_collection_index_doc
+    for collection in affected_collections:
+        update_library_collection_index_doc.delay(str(library_key), collection.key)
```
Trigger collection indexing asynchronously instead of relying on sync indexing via signals. See https://github.com/openedx/openedx-learning/pull/238/files#diff-50734dffd9d06d0a9cbc318f1b298dddd25968d064b515d716dc7a0023b7190bR660-R666

You can test the performance improvement by setting `CELERY_ALWAYS_EAGER = False` in `cms/envs/devstack.py`.

rpenido replied:
I didn't notice a performance improvement here (maybe I tested it wrong?).

This call is always in the 600ms - 2500ms range (which I agree is not good), but I don't think we should call the search tasks directly here. Maybe add another parameter to the `LIBRARY_COLLECTION_UPDATED` event to define beforehand whether the task should run sync or async.

@bradenmacdonald, can you give us your insight here? I don't want to make things complicated near the version cut.

navinkarkera replied:
> Maybe add another parameter to the LIBRARY_COLLECTION_UPDATED event to define beforehand whether the task should run sync or async.

Yes, that is better. I was probably feeling lazy to update the event as part of this ticket but seems like it is unavoidable.

navinkarkera replied:
> I didn't notice a performance improvement here (maybe I tested it wrong?).

You need to add and remove multiple collections for a component to see the difference. Also set `CELERY_ALWAYS_EAGER = False` to run celery tasks in worker.

navinkarkera replied:
@rpenido Updated `LibraryCollectionData` to include a field call `lazy` and used it here instead.

rpenido reviewed (approved):
LGTM :+1:
Thank you for your work, @navinkarkera!

- [x] I tested this using the instructions from https://github.com/openedx/frontend-app-authoring/pull/1373
- [x] I read through the code
- [x] ~I checked for accessibility issues~ N/A
- [x] Includes documentation

I added a small PR to circumvent the meilisearch bug below. It is missing some tests fixes:
- https://github.com/open-craft/edx-platform/pull/695

**Note**
We have an unused endpoint that we created beforehand for handling components/collections here: https://github.com/openedx/edx-platform/blob/15a1295777d9cb3b11f10c05fc84dc196ca3d5fc/openedx/core/djangoapps/content_libraries/views_collections.py#L194 (and some API methods too). I'm not sure if we should remove it right away. It could be helpful if we implement the "Remove from this collection" action in the three dot menu.

navinkarkera committed:
chore: add usage_key to filterable attribute

navinkarkera committed:
chore: add usage_key to filterable attribute

navinkarkera committed:
feat: add collections to component api

navinkarkera committed:
feat: add collections to component api

navinkarkera committed:
fix: lint issues

navinkarkera committed:
fix: lint issues

navinkarkera committed:
fix: remove collection-entity through model post_delete signal handler

The indexing update is already handled by m2m_changed handler.

navinkarkera committed:
fix: remove collection-entity through model post_delete signal handler

The indexing update is already handled by m2m_changed handler.

navinkarkera committed:
fix: use lazy field to control collection update task

navinkarkera committed:
fix: use lazy field to control collection update task

navinkarkera committed:
chore: temporarily point openedx-events to dev branch

navinkarkera committed:
chore: temporarily point openedx-events to dev branch

navinkarkera committed:
fix: use get_components in m2m_changed signal handler

Using `get_collection_components` does not work in case of post_remove
as the component is already removed from collection and we don't really
need additional filtering as we already have pk_set from the signal.

navinkarkera committed:
fix: use get_components in m2m_changed signal handler

Using `get_collection_components` does not work in case of post_remove
as the component is already removed from collection and we don't really
need additional filtering as we already have pk_set from the signal.

rpenido committed:
fix: circuvent meilisearch bug

rpenido commented:
Hi @navinkarkera! This is looking good!
I have a small PR about the meilisearch bug here: https://github.com/open-craft/edx-platform/pull/695
Could you check it and merge it here?

ChrisChV commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1235,6 +1256,60 @@ def update_library_collection_components(
+    assert content_library.library_key == library_key
+
+    # Note: Component.key matches its PublishableEntity.key
+    collection_qs = authoring_api.get_collections(content_library.learning_package_id).filter(
```
I think we can put this inside the `set_collections` API, check my [question](https://github.com/openedx/openedx-learning/pull/238#discussion_r1796273313)

navinkarkera replied:
Commented [here](https://github.com/openedx/openedx-learning/pull/238#discussion_r1796441135).

ChrisChV replied:
Thanks

navinkarkera committed:
Merge branch 'navin/component-collection-api' into rpenido/navin/component-collection-api-fix-meilisearch

navinkarkera committed:
Merge pull request #695 from open-craft/rpenido/navin/component-collection-api-fix-meilisearch

fix: circuvent meilisearch bug with empty objects

navinkarkera committed:
fix: update components if collection was deleted

navinkarkera committed:
refactor: rename lazy field to background

navinkarkera committed:
chore: update openedx-events version

navinkarkera committed:
chore: update openedx-learning version

ChrisChV committed:
Merge branch 'master' into navin/component-collection-api
